kernel/src/sched/mod.rs

```diff
@@ -0,0 +1,474 @@
+//! utility functions to reduce repeated code between different scheduler
+//! implementations.
+
+crate mod cooperative;
+crate mod mlfq;
+crate mod priority;
+crate mod round_robin;
+
+use core::cell::Cell;
+use core::ptr::NonNull;
+
+use crate::callback::{AppId, Callback, CallbackId};
+use crate::capabilities;
+use crate::common::cells::NumericCellExt;
+use crate::config;
+use crate::debug;
+use crate::grant::Grant;
+use crate::ipc;
+use crate::memop;
+use crate::platform::{Chip, Platform};
+use crate::process::{self, Task};
+use crate::returncode::ReturnCode;
+use crate::syscall::{ContextSwitchReason, Syscall};
+
+/// Trait which any scheduler must implement.
+pub trait Scheduler {
+    /// This function should be the last call in main.rs and should never return.
+    fn kernel_loop<P: Platform, C: Chip>(
+        &mut self,
+        platform: &P,
+        chip: &C,
+        ipc: Option<&ipc::IPC>,
+        _capability: &dyn capabilities::MainLoopCapability,
+    );
+}
+
+/// Main object for the kernel. Each board will need to create one.
+pub struct Kernel {
+    /// How many "to-do" items exist at any given time. These include
+    /// outstanding callbacks and processes in the Running state.
+    work: Cell<usize>,
+
+    /// This holds a pointer to the static array of Process pointers.
+    processes: &'static [Option<&'static dyn process::ProcessType>],
+
+    /// A counter which keeps track of how many process identifiers have been
+    /// created. This is used to create new unique identifiers for processes.
+    process_identifier_max: Cell<usize>,
+
+    /// How many grant regions have been setup. This is incremented on every
+    /// call to `create_grant()`. We need to explicitly track this so that when
+    /// processes are created they can allocated pointers for each grant.
+    grant_counter: Cell<usize>,
+
+    /// Flag to mark that grants have been finalized. This means that the kernel
+    /// cannot support creating new grants because processes have already been
+    /// created and the data structures for grants have already been
+    /// established.
+    grants_finalized: Cell<bool>,
+}
+
+impl Kernel {
+    pub fn new(processes: &'static [Option<&'static dyn process::ProcessType>]) -> Kernel {
+        Kernel {
+            work: Cell::new(0),
+            processes: processes,
+            process_identifier_max: Cell::new(0),
+            grant_counter: Cell::new(0),
+            grants_finalized: Cell::new(false),
+        }
+    }
+
+    /// Something was scheduled for a process, so there is more work to do.
+    crate fn increment_work(&self) {
+        self.work.increment();
+    }
+
+    /// Something finished for a process, so we decrement how much work there is
+    /// to do.
+    crate fn decrement_work(&self) {
+        self.work.decrement();
+    }
+
+    /// Helper function for determining if we should service processes or go to
+    /// sleep.
+    fn processes_blocked(&self) -> bool {
```

Comment:
I'm not sure if this is just a copy/paste or a new interface, but this is a confusing name. Something like `any_outstanding_work` I think would be clearer.

Reply:
this is from the old interface, and I made changes so that my PR now doesn't touch this function.